x-pack/libbeat/management/managerV2.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// Start the config manager.
+func (cm *BeatV2Manager) Start() error {
+	if !cm.Enabled() {
+		return fmt.Errorf("V2 Manager is disabled")
+	}
+	err := cm.client.Start(context.Background())
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	go cm.unitListen()
+	cm.isRunning = true
+	return nil
+}
+
+// Stop stops the current Manager and close the connection to Elastic Agent.
+func (cm *BeatV2Manager) Stop() {
+	cm.stopChan <- struct{}{}
+}
+
+// CheckRawConfig is currently not implemented for V1.
+func (cm *BeatV2Manager) CheckRawConfig(cfg *conf.C) error {
+	// This does not do anything on V1 or V2, but here we are
+	return nil
+}
+
+func (cm *BeatV2Manager) RegisterAction(action client.Action) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	cm.units[cm.mainUnit].RegisterAction(action)
+}
+
+func (cm *BeatV2Manager) UnregisterAction(action client.Action) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	cm.units[cm.mainUnit].UnregisterAction(action)
+}
+
+func (cm *BeatV2Manager) SetPayload(payload map[string]interface{}) {
+	cm.payload = payload
+}
+
+// ================================
+// Unit manager
+// ================================
+
+func (cm *BeatV2Manager) addUnit(unit *client.Unit) {
+	cm.unitsMut.Lock()
+	cm.units[unit.ID()] = unit
+	cm.unitsMut.Unlock()
+}
+
+func (cm *BeatV2Manager) getMainUnit() (*client.Unit, bool) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	if cm.mainUnit == "" {
+		return nil, false
+	}
+	return cm.units[cm.mainUnit], true
+}
+
+// We need a "main" unit that we can send updates to for the StatusReporter interface
+// the purpose of this is to just grab the first input-type unit we get and set it as the "main" unit
+func (cm *BeatV2Manager) setMainUnitValue(unit *client.Unit) {
+	cm.unitsMut.Lock()
+	defer cm.unitsMut.Unlock()
+	if cm.mainUnit == "" {
+		cm.mainUnit = unit.ID()
+	}
+}
+
+func (cm *BeatV2Manager) deleteUnit(unit *client.Unit) {
+	cm.unitsMut.Lock()
+	delete(cm.units, unit.ID())
+	cm.unitsMut.Unlock()
+}
+
+// ================================
+// Private V2 implementation
+// ================================
+
+func (cm *BeatV2Manager) unitListen() {
+	cm.logger.Debugf("Listening for agent unit changes")
+	for {
+		select {
+		// The stopChan channel comes from the Manager interface Stop() method
+		case <-cm.stopChan:
+			cm.stopBeat()
+		case change := <-cm.client.UnitChanges():
+			switch change.Type {
+			// Within the context of how we send config to beats, I'm not sure there is a difference between
+			// A unit add and a unit change, since either way we can't do much more than call the reloader
+			case client.UnitChangedAdded:
+				// At this point we also get a log level, however I'm not sure the beats core logger provides a
+				// clean way to "just" change the log level, without resetting the whole log config
+				cm.logger.Debugf("Got unit added: %s", change.Unit.ID())
+				go cm.handleUnitReload(change.Unit)
```

Comment:
Why does this need to be in goroutine? Should it also take a stopChan as an argument?

The same comment applies to `handleUnitReloaded` below.

Reply:
IIRC, at one point @blakerouse told me we need to avoid blocking that main `unitChanges` channel, hence me trying to be a little cautious and running them as separate gorutines. I think a context might be a good idea though.